datafusion/catalog/src/memory/table.rs

```diff
@@ -295,4 +304,320 @@ impl TableProvider for MemTable {
+                    BooleanArray::from(vec![true; batch.num_rows()])
+                } else {
+                    let mut combined_mask: Option<BooleanArray> = None;
+
+                    for filter_expr in &filters {
+                        let physical_expr = create_physical_expr(
+                            filter_expr,
+                            &df_schema,
+                            state.execution_props(),
+                        )?;
+
+                        let result = physical_expr.evaluate(batch)?;
+                        let array = result.into_array(batch.num_rows())?;
+                        let bool_array = array
+                            .as_any()
+                            .downcast_ref::<BooleanArray>()
+                            .ok_or_else(|| {
+                                datafusion_common::DataFusionError::Internal(
+                                    "Filter did not evaluate to boolean".to_string(),
+                                )
+                            })?
+                            .clone();
+
+                        combined_mask = Some(match combined_mask {
+                            Some(existing) => and(&existing, &bool_array)?,
+                            None => bool_array,
+                        });
+                    }
+
+                    combined_mask.unwrap_or_else(|| {
+                        BooleanArray::from(vec![true; batch.num_rows()])
+                    })
+                };
+
+                let update_count =
+                    filter_mask.iter().filter(|v| v == &Some(true)).count();
+                total_updated += update_count as u64;
+
+                if update_count == 0 {
+                    new_batches.push(batch.clone());
+                    continue;
+                }
+
+                // Normalize mask: only true (not NULL) triggers update
+                let update_mask: BooleanArray =
+                    filter_mask.iter().map(|v| Some(v == Some(true))).collect();
+
+                let mut new_columns: Vec<ArrayRef> =
+                    Vec::with_capacity(batch.num_columns());
+
+                for field in self.schema.fields() {
+                    let column_name = field.name();
+                    let original_column =
+                        batch.column_by_name(column_name).ok_or_else(|| {
+                            datafusion_common::DataFusionError::Internal(format!(
+                                "Column '{column_name}' not found in batch"
+                            ))
+                        })?;
+
+                    let new_column = if let Some(value_expr) =
+                        assignment_map.get(column_name.as_str())
+                    {
+                        let physical_expr = create_physical_expr(
+                            value_expr,
+                            &df_schema,
+                            state.execution_props(),
+                        )?;
```

Comment:
Can we create these upfront e.g. `assignment_map.map(...) -> HashMap<String, Arc<dyn PhysicalExpr>>`?